Add ability to sign in with Google Accounts

cj13579 opened the pull request:
This change adds the ability to sign in using a Google account. Users can either be added automatically upon authenticating successfully or, as right now, when added by an admin.

The new "Sign in with Google" button appears on the login screen when the `GOOGLE_CLIENT_ID` and `GOOGLE_CLIENT_SECRET` environment variables are set:

![Screenshot 2024-09-23 220403](https://github.com/user-attachments/assets/976d814f-c99b-483d-84da-a573a9110f0e)

I've linked some Google documentation in the README which describes how you can create a project and then a client ID and secret for use within this application. When you click the "Sign in with Google" button, it will redirect you off to the Google application created to authenticate. You'll see something like this:

![image](https://github.com/user-attachments/assets/d5ec43cd-c5f7-4306-8fe7-9329ee2d4cbd)

And then you'll get promoted to permit the sharing of some data:

![image](https://github.com/user-attachments/assets/af3f0aa4-5144-4e9e-bb56-4c4f0181da38)

And then you'll be rerouted back to the application and logged in(!!):

![image](https://github.com/user-attachments/assets/b4418882-4e7c-40e9-840a-70d75510fcb4)

When an administrator is adding users, there is now a new checkbox on the form to specify whether or not this user is to be an "externally authenticated user". Setting this does not set the password expiry stuff on the account:

![image](https://github.com/user-attachments/assets/c2155d0d-e94a-4abe-81ff-45b6485105eb)

You can still set a password for this user in the future if you wanted to. I haven't added anything to prevent that.

Things that could be improved/remain:

- If the admin retains control of adding users and someone authenticates successfully, no message is displayed that the "user is not found" or something. It just loops them back to the login page with no message.
- Translations for the two new properties that I have added
- Styling - it annoys me how close the text is to the logo but when I do something about it the fonts go weird.

Related to #131, maybe #26? and #25 (since this adds more dependency on Passport.js).

Wingysam commented on src/index.ts:
```diff
@@ -56,6 +57,45 @@ passport.use('local', new LocalStrategy(
+      clientSecret: process.env.GOOGLE_CLIENT_SECRET,
+      callbackURL: '/auth/google/redirect',
+    },
+    (issuer, profile, done  ) => {
```
This should use async/await and avoid `.then()`s.

Wingysam replied:
It also looks like the ESLint formatting rules aren't being applied here. Those are deprecated anyway and I need to switch the project over to prettier so I'm not too worried about that. We can merge this with whatever formatting.

cj13579 replied:
Done in https://github.com/Wingysam/Christmas-Community/pull/138/commits/173add9093dd9ad4ab64fbeb9751ead78477d44e.

Wingysam commented on src/index.ts:
```diff
@@ -56,6 +57,45 @@ passport.use('local', new LocalStrategy(
+              })          
+          }
+          else {
+            if (err.message === 'missing') return done(null, false, { message: 'Unknown user.' })
```
Ideally, this should use translations. I see that the local strategy doesn't use them either and it may not be possible in the index.ts file with my current architecture. If it can't be done without major changes, I'm okay with it.

cj13579 replied:
Localization and English translations added with https://github.com/Wingysam/Christmas-Community/pull/138/commits/736e537ee1b3c2b18bbf78b184750d4577707ac4

Wingysam commented on src/routes/adminSettings/index.js:
```diff
@@ -27,6 +27,7 @@ export default function ({ db, ensurePfp }) {
     if (!req.user.admin) return res.redirect('/')
 
     const username = req.body.newUserUsername.trim()
+    const ssoUser = req.body.isSSOUser
```
What does ssoUser do here? Also we should be storing booleans rather than strings. Please make sure not to lock in Google as the only SSO IDP. I'd also be interested in adding Discord login.

cj13579 replied:
By default, when you add a user we create an account confirmation link. This link is used by the new user to "accept" their new account and get them to set a password.

There is a change in the Admin Settings UI which adds a checkbox to the add user form. This checkbox sets this value. It is then used [here](https://github.com/cj13579/Christmas-Community/blob/google-oicd/src/routes/adminSettings/index.js#L48) to change the behavior of how we save a user. When "on", we create a user _without_ the signup token or expiry.

cj13579 replied:
I've also updated this to be a boolean: https://github.com/Wingysam/Christmas-Community/pull/138/commits/23e4a8c3310fce6ded71b53e998ea8578c3970a4.

Wingysam replied:
Okay thanks!

Wingysam commented on src/routes/google-auth/index.js:
```diff
@@ -0,0 +1,22 @@
+    passport.authenticate('google', { 
+      successRedirect: '/',
+      failureRedirect: '/login',
+      failureMessage: 'Unable to sign-in using Google'
```
Same concern with non-localized messages

cj13579 replied:
Added localization support with https://github.com/Wingysam/Christmas-Community/pull/138/commits/736e537ee1b3c2b18bbf78b184750d4577707ac4

Wingysam commented on src/routes/login/index.js:
```diff
@@ -4,12 +4,15 @@ import express from 'express'
   const router = express.Router()
 
+  var showGoogleSSO = false
+  if ( process.env.GOOGLE_CLIENT_ID && process.env.GOOGLE_CLIENT_SECRET ){ showGoogleSSO = true }
```
This should be moved to src/config/index.ts if possible.

cj13579 replied:
Fair. I've been able to remove this by recycling some existing variables: https://github.com/Wingysam/Christmas-Community/pull/138/commits/5d58cb5e3531ef5fc36618ad65429ab8612c0026.

Wingysam commented on src/index.ts:
```diff
@@ -56,6 +57,45 @@ passport.use('local', new LocalStrategy(
   }
 ))
 
+if ( process.env.GOOGLE_CLIENT_ID && process.env.GOOGLE_CLIENT_SECRET ) {
```
This should be moved to src/config/index.ts if possible

cj13579 replied:
Done as part of https://github.com/Wingysam/Christmas-Community/pull/138/commits/173add9093dd9ad4ab64fbeb9751ead78477d44e to make the Google strategy async.

Wingysam reviewed (changes requested):
Okay so this doesn't quite line up with my ideal vision for Christmas Community. For SSO in the app, I envision using something like Authentik that provides a header to tell the app what user is authenticated, rather than the app managing that state itself. However, I do see the value in not requiring an external proxy service to use SSO. I'm willing to implement the SSO natively.

I do want the IDs to be decoupled from the display names before merge. It's a bit of a heavier lift but we should store a name property separately from the _id. The _id should include the IDP. We'll need some kind of migration to append the IDP and set the name property to the ID. It should be like `mygmail@gmail.com@google` or `Sam@local`. @s in URLs are valid: [https://stackoverflow.com/a/19737890](https://stackoverflow.com/a/19737890). We could also just store pure random IDs for _id and store the IDP/ID separately.

cj13579 commented:
> Okay so this doesn't quite line up with my ideal vision for Christmas Community. For SSO in the app, I envision using something like Authentik that provides a header to tell the app what user is authenticated, rather than the app managing that state itself. However, I do see the value in not requiring an external proxy service to use SSO. I'm willing to implement the SSO natively.

I don't think that the way this has been implemented stops someone from implementing the proxy approach.

> I do want the IDs to be decoupled from the display names before merge. It's a bit of a heavier lift but we should store a name property separately from the _id. The _id should include the IDP. We'll need some kind of migration to append the IDP and set the name property to the ID. It should be like `mygmail@gmail.com@google` or `Sam@local`. @s in URLs are valid: https://stackoverflow.com/a/19737890. We could also just store pure random IDs for _id and store the IDP/ID separately.

Yeah, I agree. I was actually thinking about that this afternoon. Mostly it was because email address as the "name" of the list felt very impersonal.

Wingysam commented:
> I don't think that the way this has been implemented stops someone from implementing the proxy approach.

You're right, it just adds complexity to the app that has to be maintained alongside support for proxy-header-based auth. I'm okay with that.

> Yeah, I agree. I was actually thinking about that this afternoon. Mostly it was because email address as the "name" of the list felt very impersonal.

Maybe it would be better to do this at the same time as #23?

Wingysam commented:
What happens if someone authenticates with a different method than the account was created with? Like if I create my account with Google, then my Facebook gets compromised, can the attacker then log into my Christmas Community? We should be linking third-party unique account IDs to Christmas Community users. Ideally, I would be able to connect and remove multiple IDPs, including converting to a local account down the road if I choose to.

cj13579 commented:
> What happens if someone authenticates with a different method than the account was created with? Like if I create my account with Google, then my Facebook gets compromised, can the attacker then log into my Christmas Community? We should be linking third-party unique account IDs to Christmas Community users. Ideally, I would be able to connect and remove multiple IDPs, including converting to a local account down the road if I choose to.

Yeah, this makes sense. I'll look into that...

cj13579 commented:
Alright, @Wingysam I think that this might be more like what you are thinking. I have tweaked a number of bits:

- Removed the capability to effectively "sign-up" if you successfully login
- Added a "link account" capability to the Profile screen

An admin will add a user just as they do today, that user will still need to confirm their account and set a password. After they have done this, they can go to their profile and link their Google account:

![image](https://github.com/user-attachments/assets/e5217c6d-b509-49cc-afff-618bc56a791e)

Once you have linked your account, you can unlink it by going back to the _Security_ section of your profile:

![image](https://github.com/user-attachments/assets/1a2e42dc-c8c0-4c19-9a74-be78de383dcd)

When you have a linked account, you can login using either your local username/password _or_ your Google account. In this way, we can add in support for Discord, Facebook, Microsoft, Apple, etc. and people can link as many or as few as they'd like.

Wingysam commented:
I think when the user opens the magic link after the admin creates their account, they should get something like

> Welcome to $SITENAME! How would you like to log in going forward?
>
> \[Username and password\] \[Google\]

Then if they have multiple forms of auth, they may remove one. If you only have Google or only have password, you must add another form before removing the first.

cj13579 commented:
That feels like something that could be added later, IMO.

cj13579 commented:
@Wingysam I've done a bit more exploring on this:

> I think when the user opens the magic link after the admin creates their account, they should get something like

It's harder than I thought it would be. Thinking about the things you said though: If they sign-up with a password, then they will always have that mechanism as an option. Adding additional methods can be added/removed as and when.

Ideally, I'd like to get this in soon since Christmas is fast approaching. I think once Google is in, I could pretty quickly add Microsoft and Discord. I'm just concerned about merge conflicts if I start on those now before this gets in. Selfishly, those 3 cover the needs of my family and friends!

Let me know if you're not keen on this approach though. I am happy to just keep some of these mods in a fork if I'm heading on a path you're not comfortable with.

Wingysam commented:
I'm not too concerned about getting more IDPs in immediately. User/pass + Google is good enough for now. It just needs to be implemented like
```json
{
  "_id": "Wingy",
  "password": "<hash>",
  "oauthConnections": {
    "google": "<google user id>"
  }
}
```
I'm planning on migrating to SQLite eventually. At that point we'll have a table of login connections relating to users. Until then, we can just iterate through every user for matching Google accounts. 

We will need to deal with multiple users paired to one Google account by either not letting you connect a Google account if someone is already using it or a Netflix "Who's watching?"-style prompt.

cj13579 commented:
@Wingysam fair enough. Hopefully the last couple of changes in https://github.com/Wingysam/Christmas-Community/pull/138/commits/f1be4a8f0defd62217b59ef36c75056dc285b245 (`oauthConnections.google` > `idpGoogle`) and https://github.com/Wingysam/Christmas-Community/pull/138/commits/a41ce158c0b1809278756289586fb831da728acd (display an error -like below- if a Google account is already linked to an existing site account) is enough to get this PR over the line:

![image](https://github.com/user-attachments/assets/8aed4a41-88c6-45ad-a700-0c8f412c5c73)

cj13579 commented:
@Wingysam I found one more thing with this which I have pushed a commit for. I saw http scheme issue (http vs https) when running in a docker image. I run my docker images behind nginx and I terminate my TLS there. Previously, I had hardcoded relative URI's but, I assume because the Express server is not running in HTTPS mode, the redirect links were being created with HTTP not HTTPS (required by Google and others).

This latest change allows you to set an environment variable to specify the full URLs for the redirects needed.

As an aside, I had real issues getting the docker stuff building and had to make a few changes. I expect you know about these though.

Wingysam commented on src/index.ts:
```diff
@@ -96,21 +96,29 @@ if ( config.googleSSOEnabled ) {
+      } else {
+        try {
+          const doc = await db.get(req.session.passport.user)
+          if (doc.oauthConnections) {
```
My preference is for this to be done like this:
```js
doc.oauthConnections ??= {} // handle migration from 1.35 and earlier releases
doc.oauthConnections.google = googleId // TODO: call the variable googleId rather than googleEmail
```

There's also an argument to be made for the following, since it doesn't use newish/obscure JS features. I prefer the above though.
```js
if (typeof doc.oauthConnections === 'undefined') doc.oauthConnections = {} // handle migration from 1.35 and earlier releases
doc.oauthConnections.google = googleId
```

cj13579 replied:
Agree. Updated with https://github.com/Wingysam/Christmas-Community/pull/138/commits/da1375887114c48ea581fb5a91d46f3891d563b4

Wingysam commented on README.md:
```diff
@@ -132,6 +132,9 @@ MARKDOWN=false
 # GOOGLE_CLIENT_ID=abc123
 # GOOGLE_CLIENT_SECRET=123abc
+# When running in docker, you may face issues with the protocol or URL that is redirecting the call, override that by specifying the full redirect URL in the below variables
+# GOOGLE_SIGNIN_REDIRECT=https://wishlist.example.com/auth/google/redirect
```
I'd prefer to have the user set the base URL of the site (similar to the ROOT_PATH we have now) and extrapolate from there. There's some complexity with that, but changing how variables like this work in the future is painful because it's a breaking change.

Maybe a ROOT_URL variable that is required if specifying GOOGLE_CLIENT_ID and GOOGLE_CLIENT_SECRET?

So I could do something like ROOT_URL=https://wishlist.example.com/subpath and it would automatically overwrite base with its path? So something like this:
```ts
function appendSlash(path: string) {
  if (path.endsWith('/')) return path
  return path + '/'
}

...
rootUrl: appendSlash(process.env.ROOT_URL ?? process.env.ROOT_PATH ?? '/')
...

// The base path is used in HTML templates rather than the fully qualified path, mostly for legacy reasons. It also has the following advantages:
//  * It saves a negligible amount of bandwidth.
//  * It allows the admin to deploy to multiple subdomains. This isn't really supported, but it's nice if the admin migrates domains on the same path (typically /), and can keep the old one around because users won't migrate immediately.
// This used to be set with ROOT_PATH, but is now calculated based on ROOT_URL, which falls back to the former if unset.
// rootUrl may be a fully-qualified `https://domain/path/`-style value, or may simply be `/`, or a subpath.
try {
  config.base = new URL(config.rootUrl).pathname
} catch {
  config.base = config.rootUrl
}
```

cj13579 replied:
@Wingysam, do you want to implement this last piece? Sounds like you have it in your head what you're looking for.

Wingysam replied:
Sure, I'll do that part!

Wingysam commented on src/index.ts:
```diff
@@ -45,17 +46,85 @@ passport.use('local', new LocalStrategy(
+      callbackURL: config.googleSignInRedirect,
+    },
+    async (issuer, profile, done) => {
+      const username = profile.emails[0].value.trim();
```
This should use `profile.id` rather than the first email, because the emails associated with a Google account can change.

cj13579 replied:
Resolved with https://github.com/Wingysam/Christmas-Community/pull/138/commits/607b7887945799def64dee7415b2b739b82f4151

Wingysam commented on src/index.ts:
```diff
@@ -45,17 +46,85 @@ passport.use('local', new LocalStrategy(
+    passReqToCallback: true
+  },
+    async (req, issuer, profile, done) => {
+      const googleEmail = profile.emails[0].value.trim();  // Get Google email
```
Same concern re. emails vs ids.

cj13579 replied:
Resolved with https://github.com/Wingysam/Christmas-Community/pull/138/commits/607b7887945799def64dee7415b2b739b82f4151

Wingysam commented on src/routes/google-auth/index.js:
```diff
@@ -0,0 +1,46 @@
+  const router = express.Router()
+
+  router.get('/', passport.authenticate('google-login', {
+    scope: ['openid', 'profile', 'email']
```
Shouldn't need the email scope once you switch to using ids.

cj13579 replied:
Resolved with https://github.com/Wingysam/Christmas-Community/pull/138/commits/607b7887945799def64dee7415b2b739b82f4151

Wingysam commented on src/routes/google-auth/index.js:
```diff
@@ -0,0 +1,46 @@
+  );
+
+  router.get('/link', passport.authenticate('google-link', {
+    scope: ['email']
```
Same comment re. email scope.

cj13579 replied:
Resolved with https://github.com/Wingysam/Christmas-Community/pull/138/commits/607b7887945799def64dee7415b2b739b82f4151

Wingysam commented on src/views/profile.pug:
```diff
@@ -57,3 +57,15 @@ block content
     span= lang('PROFILE_SECURITY_CHANGE_PASSWORD')
+  | &nbsp;
+  if _CC.config.googleSSOEnabled
+    if  req.user.oauthConnections && req.user.oauthConnections.google
```
There's an extra space at the beginning of this if statement.

cj13579 replied:
Resolved with https://github.com/Wingysam/Christmas-Community/pull/138/commits/599e27d877eb261d42b80de95c05c485bdf67333

Wingysam commented:
First off, I want to say thank you for putting in so much work to get this off the ground. Sorry my response times are sporadic. I've been busy with college and work without any time for hobbies lately. 

This is looking close to ready, but has a couple of changes that wouldn't be possible to make post-merge without breaking backwards-compatibility, so I'd like those fixed first. If you don't have the time to do that, please let me know and I'll see if I can finish it up next week.

Wingysam commented:
> As an aside, I had real issues getting the docker stuff building and had to make a few changes. I expect you know about these though.

I believe these were fixed by #145, I didn't update the Dockerfile when I migrated the project to TypeScript and haven't tried to make a release since then. I'm hoping to make a release as soon as this is merged.

cj13579 commented:
> I believe these were fixed by #145, I didn't update the Dockerfile when I migrated the project to TypeScript and haven't tried to make a release since then. I'm hoping to make a release as soon as this is merged.

That didn't work for me since there is still a double ampersand missing on line 17. I also got a bunch of tsc "module not found" type things. The dockerfile that I am running with for now is [this](https://github.com/cj13579/Christmas-Community/blob/wishlists-2024/Dockerfile). I think it increases the image size a bunch but that's not the end of the world for me.

Wingysam commented:
I see that the Dockerfile still definitely doesn't work. I'll experiment locally with making one that works properly and efficiently.
